apps/anoma_node/lib/node/transaction/ordering.ex

```diff
@@ -357,20 +579,483 @@ defmodule Anoma.Node.Transaction.Ordering do
+    state_with_watermarks =
+      recalculate_and_update_watermarks_for_keys(
+        keys_to_update,
+        state_without_failed_tx
+      )
+
+    try_finalize_transaction(tx_id, state_with_watermarks)
+  end
+
+  @spec block_reservations(
+          String.t(),
+          binary(),
+          [{:read | :write, binary()}]
+        ) :: :ok
+  defp block_reservations(node_id, tx_id, reservations) do
+    receive do
+      %EventBroker.Event{
+        body: %Node.Event{body: %__MODULE__.OrderEvent{tx_id: ^tx_id}}
+      } ->
+        request_reservations(node_id, tx_id, reservations)
+
+      _ ->
+        IO.puts("this should be unreachable")
+    end
+
+    EventBroker.unsubscribe_me([
+      Node.Event.node_filter(node_id),
+      this_module_filter(),
+      tx_id_filter(tx_id)
+    ])
+  end
+
+  # Helper to release reservations on Shards
+  @spec release_reservations(%{binary() => pid()}, integer()) :: :ok
+  defp release_reservations(shard_pids, height) do
+    Enum.each(shard_pids, fn {_key, shard_pid} ->
+      # Cast should be okay for unreserve as it's best-effort rollback
+      Shard.unreserve(shard_pid, height)
+    end)
+  end
+
+  # Recalculates watermarks for a list of keys and sends updates to shards if they advance.
+  @spec recalculate_and_update_watermarks_for_keys([binary()], t()) :: t()
+  defp recalculate_and_update_watermarks_for_keys(keys, state) do
+    Enum.reduce(keys |> Enum.uniq(), state, fn key, acc_state ->
+      recalculate_and_update_watermarks_for_key(key, acc_state)
+    end)
+  end
+
+  # Recalculates watermarks for a single key and sends updates to shards if they advance.
+  @spec recalculate_and_update_watermarks_for_key(binary(), t()) :: t()
+  defp recalculate_and_update_watermarks_for_key(key, state) do
+    current_wms =
+      Map.get(state.watermark_state, key, %{
+        read_watermark: -1,
+        write_watermark: -1
+      })
+
+    new_read_wm = calculate_watermark(key, :read, state)
+    new_write_wm = calculate_watermark(key, :write, state)
+
+    state_after_update = state
+
+    # Check and update read watermark
+    state_after_update =
+      if new_read_wm > current_wms.read_watermark do
+        case get_shard_pid(key, state) do
+          {:ok, shard_pid} ->
+            send(shard_pid, {:read_watermark_advanced, key, new_read_wm})
+
+          {:error, reason} ->
+            Logger.error(
+              "Failed to get shard PID for key #{inspect(key)} when advancing read watermark: #{inspect(reason)}"
+            )
+
+            # Continue without sending update
+        end
+
+        updated_key_wms = Map.put(current_wms, :read_watermark, new_read_wm)
+
+        %{
+          state_after_update
+          | watermark_state:
+              Map.put(
+                state_after_update.watermark_state,
+                key,
+                updated_key_wms
+              )
+        }
+      else
+        # No change
+        state_after_update
+      end
```

Comment:
ditto here.... wait isn't this the same logic... this should be abstracted, this is all too manual